xray/plot/facetgrid.py

```diff
@@ -0,0 +1,380 @@
+
+        """
+
+        import matplotlib.pyplot as plt
+
+        # Handle corner case of nonunique coordinates
+        rep_col = col is not None and not data[col].to_index().is_unique
+        rep_row = row is not None and not data[row].to_index().is_unique
+        if rep_col or rep_row:
+            raise ValueError('Coordinates used for faceting cannot '
+                             'contain repeated (nonunique) values.')
+
+        # self._single_group is the grouping variable, if there is exactly one
+        if col and row:
+            self._single_group = False
+            self._nrow = len(data[row])
+            self._ncol = len(data[col])
+            self.nfacet = self._nrow * self._ncol
+            if col_wrap is not None:
+                warnings.warn('Ignoring col_wrap since both col and row '
+                              'were passed')
+        elif row and not col:
+            self._single_group = row
+        elif not row and col:
+            self._single_group = col
+        else:
+            raise ValueError(
+                'Pass a coordinate name as an argument for row or col')
+
+        # Compute grid shape
+        if self._single_group:
+            self.nfacet = len(data[self._single_group])
+            if col:
+                # idea - could add heuristic for nice shapes like 3x4
+                self._ncol = self.nfacet
+            if row:
+                self._ncol = 1
+            if col_wrap is not None:
+                # Overrides previous settings
+                self._ncol = col_wrap
+            self._nrow = int(np.ceil(self.nfacet / self._ncol))
+
+        # Calculate the base figure size with extra horizontal space for a
+        # colorbar
+        self._cbar_space = 1
+        figsize = (self._ncol * size * aspect +
+                   self._cbar_space, self._nrow * size)
+
+        self.fig, self.axes = plt.subplots(self._nrow, self._ncol,
+                                           sharex=True, sharey=True,
+                                           squeeze=False, figsize=figsize)
+
+        # Set up the lists of names for the row and column facet variables
+        col_names = list(data[col].values) if col else []
+        row_names = list(data[row].values) if row else []
+
+        if self._single_group:
+            full = [{self._single_group: x} for x in
+                    data[self._single_group].values]
+            empty = [None for x in range(self._nrow * self._ncol - len(full))]
+            name_dicts = full + empty
+        else:
+            rowcols = itertools.product(row_names, col_names)
+            name_dicts = [{row: r, col: c} for r, c in rowcols]
+
+        self.name_dicts = np.array(name_dicts).reshape(self._nrow, self._ncol)
+
+        self.row_names = row_names
+        self.col_names = col_names
+        self.data = data
+        self.row = row
+        self.col = col
+        self.col_wrap = col_wrap
+
+        # Next the private variables
```

Comment:
These variables are all duplicated between private versions (below) and public versions (above). For now, let's stick with the conservative choice of keeping them private like seaborn?